graphql_compiler/cost_estimation/analysis.py

```diff
@@ -0,0 +1,238 @@
+) -> Interval[Any]:
+    lower_bound = None
+    upper_bound = None
+    if interval.lower_bound is not None:
+        lower_bound = convert_int_to_field_value(
+            schema_info, vertex_type, field, interval.lower_bound
+        )
+    if interval.upper_bound is not None:
+        upper_bound = convert_int_to_field_value(
+            schema_info, vertex_type, field, interval.upper_bound
+        )
+    return Interval(lower_bound, upper_bound)
+
+
+def get_field_value_intervals(
+    schema_info: QueryPlanningSchemaInfo,
+    query_metadata: QueryMetadataTable,
+    parameters: Dict[str, Any],
+) -> Dict[PropertyPath, Interval[Any]]:
+    """Map each PropertyPath in the query to its field value interval if it can be computed.
+
+    Args:
+        - schema_info: QueryPlanningSchemaInfo
+        - query_metadata: a representation of the query
+        - parameters: the parameters used for the query
+
+    Returns:
+        Dict mapping some PropertyPaths to the interval of filtered values at that field.
+    """
+    field_value_intervals = {}
+    for location, location_info in query_metadata.registered_locations:
+        filter_infos = query_metadata.get_filter_infos(location)
+        vertex_type_name = location_info.type.name
+
+        # Group filters by field
+        single_field_filters: Dict[str, List[FilterInfo]] = {}
+        for filter_info in filter_infos:
+            if len(filter_info.fields) == 0:
+                raise AssertionError(f"Got filter on 0 fields {filter_info} on {vertex_type_name}")
+            elif len(filter_info.fields) == 1:
+                single_field_filters.setdefault(filter_info.fields[0], []).append(filter_info)
+            else:
+                pass  # We don't do anything for multi-field filters yet
+
+        # Find field_value_interval for each field
+        for field_name, filters_on_field in single_field_filters.items():
+            filters_on_field = [
+                filter_info for filter_info in filter_infos if filter_info.fields == (field_name,)
+            ]
+
+            integer_interval = get_integer_interval_for_filters_on_field(
+                schema_info, filters_on_field, vertex_type_name, field_name, parameters
+            )
+            field_value_interval = _convert_int_interval_to_field_value_interval(
+                schema_info, vertex_type_name, field_name, integer_interval
+            )
+
+            field_value_intervals[(location.query_path, field_name)] = field_value_interval
+    return field_value_intervals
+
+
+def get_distinct_result_set_estimates(
+    schema_info: QueryPlanningSchemaInfo,
+    query_metadata: QueryMetadataTable,
+    parameters: Dict[str, Any],
+) -> Dict[VertexPath, float]:
+    """Map each VertexPath in the query to its distinct result set estimate.
+
+    The distinct result set estimate for a query node is the expected number of different
+    vertices that will appear under it in the result of the query.
```

Comment:
How about:
`The distinct result set estimate for vertex query node is the expected number of different instances of the vertex type that will appear in the result set of the query. For instance, suppose a query that included an edge traversal from A  to B that also included a unique filter on A. In this case, the distinct result estimate for A is 1 even though the cardinality of the result set might be quite large.`